File: Algorithm-Sources/Algospot/알고리즘 문제 해결 전략/Chapter 8/pi.cpp
#include<stdio.h>
#include<stdlib.h>
#include<memory.h>
#include<string.h>
#include<iostream>
#include<algorithm>
using namespace std;

void getDiff(char *s, int l, int *rtn) {
	int off = l-3;
	if(s[off] == s[off+1] && s[off] == s[off+2])
		rtn[0] = 1;
	else if(s[off] - s[off+1] == s[off+1] - s[off+2])
		if(abs(s[off] - s[off+1]) == 1)
			rtn[0] = 2;
		else
			rtn[0] = 5;
	else if(s[off] == s[off+2])
		rtn[0] = 4;
	else
		rtn[0] = 10;
	if(l > 3) {
		if(rtn[0] == 1 && s[off-1] == s[off])	rtn[1] = 1;
		else if((rtn[0] == 2 || rtn[0] == 5) && s[off-1] - s[off] == s[off] - s[off+1]) rtn[1] = rtn[0];
		else if(rtn[0] == 4 && s[off-1] == s[off+1]) rtn[1] = 4;
		else rtn[1] = 10;
	}
	if(l == 5) {
		off--;
		if(rtn[1] == 1 && s[off-1] == s[off])	rtn[2] = 1;
		else if((rtn[1] == 2 || rtn[1] == 5) && s[off-1] - s[off] == s[off] - s[off+1]) rtn[2] = rtn[1];
		else if(rtn[1] == 4 && s[off-1] == s[off+1]) rtn[2] = 4;
		else rtn[2] = 10;
	}
}

int main() {
//	freopen("input.in", "r", stdin);
	int caseNum;
	scanf("%d", &caseNum);
	
	char s[10001];
	while(caseNum--) {
		scanf("%s", s);
		int d[10001];
		memset(d, 0x0f, sizeof(d));
		
		int len = (int)strlen(s);
		int rtn[3];
		d[0] = 0;
		getDiff(s, 3, rtn);
		d[3] = rtn[0];
		getDiff(s, 4, rtn);
		d[4] = rtn[1];
		for(int i = 5;i<=len;i++) {
			getDiff(s+i-5, 5, rtn);
			d[i] = min(d[i], d[i-5] + rtn[2]);
			d[i] = min(d[i], d[i-4] + rtn[1]);
			d[i] = min(d[i], d[i-3] + rtn[0]);
		}
		printf("%d\n", d[len]);
	}
}

File: Algorithm-Sources/Algospot/알고리즘 문제 해결 전략/Chapter 8/pi.py
import sys
rl = lambda : sys.stdin.readline()

caseNum = int(rl())

def diff(l):
	l = list(reversed(l))
	rtn = [10, 10, 10]
	if l[0] == l[1] and l[1] == l[2]:
		rtn[0] = 1
		if len(l) > 3 and l[2] == l[3]:
			rtn[1] = 1
			if len(l) == 5 and l[3] == l[4]:
				rtn[2] = 1
	elif l[0] == l[1]-1 and l[1] == l[2]-1:
		rtn[0] = 2
		if len(l) > 3 and l[2] == l[3]-1:
			rtn[1] = 2
			if len(l) == 5 and l[3] == l[4]-1:
				rtn[2] = 2
	elif l[0] == l[1]+1 and l[1] == l[2]+1:
		rtn[0] = 2
		if len(l) > 3 and l[2] == l[3]+1:
			rtn[1] = 2
			if len(l) == 5 and l[3] == l[4]+1:
				rtn[2] = 2
	elif l[0] == l[2]:
		rtn[0] = 4
		if len(l) > 3 and l[1] == l[3]:
			rtn[1] = 4
			if len(l) == 5 and l[2] == l[4]:
				rtn[2] = 4
	elif l[0] - l[1] == l[1] - l[2]:
		rtn[0] = 5
		if len(l) > 3 and l[2] - l[3] == l[1] - l[2]:
			rtn[1] = 5
			if len(l) == 5 and l[2] - l[3] == l[3] - l[4]:
				rtn[2] = 5
	return rtn

for c in range(caseNum):
	ar = [int(i) for i in rl().strip()]
	d = [0] * len(ar)
	d[0:2] = [-1]*2
	d[2:5] = diff(list(reversed(ar[:5])))
	#	print d
	for i in range(5, len(ar)):
		#	print ar[i-4:i+1], diff(ar[i-4:i+1])
		#print d[i-5:i-2], [d[i-5:i-2][k] + diff(ar[i-4:i+1])[k] for k in range(3) if d[i-5:i-2][k] != -1]
		rtn = diff(ar[i-4:i+1])
		d[i] = 9999999999
		#		if d[i] > d[i-5] + rtn[0] and d[i-5] != -1: d[i] = d[i-5] + rtn[0]
		#if d[i] > d[i-4] + rtn[1] and d[i-4] != -1: d[i] = d[i-4] + rtn[1]
	#if d[i] > d[i-3] + rtn[2] and d[i-3] != -1: d[i] = d[i-3] + rtn[2]

	#print d
	print d[len(ar)-1]

# python으로 해서 그런가 1초안에 동작 안한다 -_-;